prover/src/lib.rs

```diff
@@ -192,44 +194,73 @@ pub trait Prover {
     // --------------------------------------------------------------------------------------------
 
     /// Returns a STARK proof attesting to a correct execution of a computation defined by the
-    /// provided trace.
+    /// provided trace. The trace must not contain an auxiliary trace segment.
     ///
     /// The returned [StarkProof] attests that the specified `trace` is a valid execution trace of
     /// the computation described by [Self::Air](Prover::Air) and generated using some set of
     /// secret and public inputs. Public inputs must match the value returned from
     /// [Self::get_pub_inputs()](Prover::get_pub_inputs) for the provided trace.
     fn prove(&self, trace: Self::Trace) -> Result<StarkProof, ProverError> {
-        // figure out which version of the generic proof generation procedure to run. this is a sort
-        // of static dispatch for selecting two generic parameter: extension field and hash
-        // function.
-        match self.options().field_extension() {
-            FieldExtension::None => self.generate_proof::<Self::BaseField>(trace),
-            FieldExtension::Quadratic => {
-                if !<QuadExtension<Self::BaseField>>::is_supported() {
-                    return Err(ProverError::UnsupportedFieldExtension(2));
-                }
-                self.generate_proof::<QuadExtension<Self::BaseField>>(trace)
-            }
-            FieldExtension::Cubic => {
-                if !<CubeExtension<Self::BaseField>>::is_supported() {
-                    return Err(ProverError::UnsupportedFieldExtension(3));
-                }
-                self.generate_proof::<CubeExtension<Self::BaseField>>(trace)
-            }
-        }
-    }
+        // FIXME: the `self.options().field_extension` is now irrelevant and could be inconsistent
+        // with `E`.
+        assert!(!trace.info().is_multi_segment());
```

Comment:
A big change in this PR is that the field extension is now known at compile-time. If the trace has an auxiliary segment, the caller must call `prove_with_aux_trace()`, which takes a generic `E`.

The `ProofOptions::field_extension` ultimately is limiting and [results in this "hack"](https://github.com/facebook/winterfell/blob/09525751727a283dfbf5e569a09909b82380e059/prover/src/lib.rs#L206-L219) (removed in this PR). Since completely removing `FieldExtension` is a larger change, we should do it in a subsequent cleanup PR.

Reply:
Sometimes we do want to pick field extension dynamically (e.g., degree 2 extension of the `f64` field can be used for ~100 bit security, while extension 3 is needed for 128 bit security).

How would we achieve this dynamic selection with this structure?

Reply:
Could you give an example of a use case where this needs to be dynamic?

Could we not solve this by including type aliases for each field, e.g.

```rs
// f64 field

pub type ExtensionField100BitSecurity = QuadExtension<BaseElement>;
pub type ExtensionField128BitSecurity = CubeExtension<BaseElement>;
```